doc: add guide for backporting PRs

evanlucas opened the pull request:
This guide should help answer questions for contributors
that are not familiar with the backport process.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
doc

addaleax commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+
+Each release line has a staging branch that the releaser will use a scratch pad
+while preparing a release. The branch name is formatted as follows:
+`vn.x-staging` where `n` is the major release number.
```
Just a suggestion but how about a capital N? `vN.x-staging` might be a bit clearer.

gibfahn commented:
Does this cover the same ground as https://github.com/nodejs/node/pull/10343?

If so I guess we should decide whether we want this as a separate guide linked to from the COLLABORATOR_GUIDE (i.e. this PR), or just a note at the bottom of the COLLABORATOR_GUIDE https://github.com/nodejs/node/pull/10343.

I'm undecided.

evanlucas commented:
@gibfahn oh wow, thanks. I totally missed that one

jasnell commented:
There's likely value in separating it out

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+
+## Staging branches
+
+Each release line has a staging branch that the releaser will use a scratch pad
```
"as a scratch"

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+If a cherry-pick does not landing cleanly on the staging branch, the releaser
+will mark the pull request with a particular label for that release line,
+specifying to our tooling that this pull request should not be included. The
+releaser will then add a comment that a backport is needed.
```
what is the label? the labelelling has to distinguish things that should be backported (but aren't landing cleanly) from things that should not be backported (clean cherry-pick or not)

gibfahn replied:
Related: https://github.com/nodejs/github-bot/issues/120

evanlucas replied:
Yea I wasn't quite sure how to word it being as the label is different for each release line.

I agree that we should have 1 label for something that doesn't land and another for 1 that shouldn't land. That being said, I think the reasoning for not having them is that we should land everything in master that isn't a breaking change and a breaking change will be marked as semver major.

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+## What needs to be backported?
+
+When a release is being prepared, the releaser attempts to cherry-pick a
+certain set of commits from the master branch to the release staging branch.
```
"certain" is vague, but elaborated on below, maybe state that criteria for consideration depends on the the target version (current or LTS)?

sam-github replied:
actually, not elaborated on below, I don't think.

evanlucas replied:
Yea, I hear ya. It is a bit difficult to find the middle ground between spilling my guts on cutting releases vs the actual intention of the guide. I'll improve this.

gibfahn replied:
I think ideally this would link to the part of [releases.md](https://github.com/nodejs/node/blob/master/doc/releases.md) which talks about the criteria, not sure if there is currently anything in there to link to though.

evanlucas replied:
There is not anything in the releases.md document regarding this. I'll work on documenting that more specifically as well.

rvagg replied:
I don't think that level of information belongs in this doc, if we can actually figure out what "certain" means then it belongs elsewhere anyway.

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+
+## What can be backported?
+
+The "Current" release line (currently v7.x) is much more lenient in what can
```
lenient than what?

sam-github replied:
shouldn't semver-major be mentioned here somewhere?

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+
+## How to submit a backport pull request?
+
+For these steps, let's assume that a backport is needed for the v7.x release
```
This is unusual, isn't it? that a backport is needed from master to Current? Shouldn't master always pick clean to Current?

evanlucas replied:
not necessarily. Master takes semver-major changes. I see a lot of conflicts from changes that go in after a major change so they have to be backported quite a bit in that case.

sam-github commented on doc/guides/backporting-to-release-lines.md:
````diff
@@ -0,0 +1,81 @@
+git checkout -b backport-10157
+```
+
+*Note: the branch name does not have to have the pull request number in it.*
````
Recommending a consistent branch naming is a good thing: `backport-<PR#>-to-<target node version>`

evanlucas replied:
I like it. Will update

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+* After creating the branch, apply the changes to the branch.
+* The commit message should be identical to the commit message on the master
+  branch, unless the commit has to be different due to dependencies that are
+  not present in the targeted release line.
```
what do you mean by "deps not present"? If its a manual backport, the code is changing, sometimes a lot, should the commit message still be exactly the same?

evanlucas replied:
That's a fair point. I guess the answer in that case would be "it depends". I don't think we should have identical commit messages if they are incorrect, but I do think we need a better way to show, hey this is a backport commit (like the v8 backports).

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+### Helpful Hints
+
+* Please include the backport target in the pull request title.
+  * Ex. `process: improve performance of nextTick (v4.x backport)`
```
Instead of an example, I suggest we be prescriptive. I've been doing "v4.x backport: proces: improve...", I thought that was requested, but I'm happy to do above instead.

gibfahn replied:
I'd rather we put "v4.x backport" at the beginning makes it stand out more.

+1 for specifying exactly.

evanlucas replied:
that works for me. I added this so we could actually come up with a specified way that we request (and can edit to make sure is followed). I would rather we be opinionated on this. It makes looking for backports easier IO

sam-github commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+A good example of a commit that has been backported is
+https://github.com/nodejs/node/commit/2d23562588eebaf9a2a3d3987fe6e4ef63ab9500.
+
+* Original Pull Request: https://github.com/nodejs/node/pull/10157
```
neither the commit above, nor the PR it links to, has these two lines in the commit message or PR description, can you point to where they should be? In the commit message when it lands, the backport description?

Maybe our process should add another `PR-URL`-like piece of metadata for commits that had to be reworked and PRed as a backport because they didn't land clean? The current metadata seems like a bit of a lie, the reviewers never saw the particular code diff that is landing in the backport.

gibfahn replied:
I thought that the procedure was to get LGTMs on the backport PR, and include those (not the LGTMs from the original PR) when landing.

evanlucas replied:
I do think we need to come up with an additional piece of metadata for the original PR-URL. Part of the reason for opening this is so we can get to the point where *everyone* knows how to handle these, because right now, things are kind of all over the place. +1 for including the reviewers from the backport PR vs the original PR

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+
+When a release is being prepared, the releaser attempts to cherry-pick a
+certain set of commits from the master branch to the release staging branch.
+If a cherry-pick does not landing cleanly on the staging branch, the releaser
```
`landing`->`land`

Also, do we do three-way merging (`git am -3`)? If so it might be worth specifying that.

evanlucas replied:
no, we only cherry-pick (or I only cherry-pick). /cc @nodejs/release do any of yall actually fixup conflicts when cherry-picking?

gibfahn replied:
I'm pretty sure I got this (in my `~/.gitconfig`) from @MylesBorins :

```bash
  pa = "!curl -L $1.patch | git am --whitespace=fix #" # Patch Github PR
  pap = "!curl -L $1.patch | git am -3 --whitespace=fix #" # Please patch it
```

evanlucas replied:
Yes, when landing a commit for pushing to master, we use `git am`. For backporting, the goal is to get the commit unmodified, so we use `git cherry-pick <sha>`.

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+# Assuming your fork of Node.js is checked out in $NODE_DIR,
+# the origin remote points to your fork, and the upstream remote points
+# to git://github.com/nodejs/node
+cd $NODE_DIR
```
Might be easier to just say something like:

```bash
git clone -o upstream https://github.com/nodejs/node.git && cd node
git remote add fork git@github.com:$USER/node.git
git fetch --all
```

I'd find it easier to understand, and it doubles as a guide for someone not sure about setting up multiple remotes.

evanlucas replied:
sure, although I would think that `origin` is more common than `fork`

gibfahn replied:
I like to be specific in instructions, origin can end up being `upstream` or `fork`.

evanlucas replied:
I guess I'm a lot more used to https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-1-fork

gibfahn replied:
Okay fair enough, makes sense to keep it consistent then.

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+
+*Note: the branch name does not have to have the pull request number in it.*
+
+* After creating the branch, apply the changes to the branch.
```
Maybe specify cherry-picking the commit across?

```bash
git cherry-pick $SHA # Use your commit hash
```

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,81 @@
+cherry-picked cleanly from master, then nothing needs to be done. If not, a
+backport pull request will need to be submitted.
+
+## How to submit a backport pull request?
```
Don't need the `?`

gibfahn commented:
We should definitely link to this from the COLLABORATOR_GUIDE.

evanlucas commented:
Ok, updated. PTAL. I went ahead and added to the COLLABORATOR_GUIDE as well. I removed an example commit since I (unfortunately) couldn't find one that actually used the backport PR-URL instead of the original. Thanks!

mhdawson commented on doc/guides/backporting-to-release-lines.md:
````diff
@@ -0,0 +1,84 @@
+```shell
+git cherry-pick $SHA # Use your commit hash
+```
+
````
At this point do we need to say what you should expect to see and how to fix it up? I assume that following this guide is only needed if the cherry-pick is going to fail (I'm also assuming that $SHA is for the original commit that went into master)

evanlucas replied:
I'll clarify. Yes, this guide is for when the releaser cannot cleanly cherry-pick the commit and requests a backport.

joyeecheung commented on doc/guides/backporting-to-release-lines.md:
````diff
@@ -0,0 +1,84 @@
+```
+
+* The commit message should be as close as possible to the commit message on the
+  master branch, unless the commit has to be different due to dependencies that
````
If I understand correctly, this means the metadata(PR urls, reviewers, etc) should remain the same as the original commit?

evanlucas replied:
In the case of a backport, I *don't* think it should. I'll clarify that.

sam-github commented:
Maybe we should have explicit meta-data on backport commits, like `Backport: <URL to PR being backported>`?

gibfahn commented:
Having another line seems like a good idea, so @sam-github are you suggesting that if we have:

```
PR-URL: url1
Backport: url2
```

then `PR-URL: url1` is the original PR, and `Backport: url2` is the `vN.x Backport` PR (not the other way around)? I guess it might make sense to keep `PR-URL:` the same across backports.

evanlucas commented:
@sam-github @gibfahn I agree. I went ahead and added it in. PTAL. Thanks!

targos commented:
In the metadata of the backport commit, who should be displayed as `Reviewed-By`?
- Reviewers of the original PR?
- Reviewers of the backport PR?
- Both?

evanlucas commented:
@targos good point. IMO, if a backport PR is necessary, it should go through the standard process of reviewers. I think the reviewers from the backport PR should be listed, not the reviewers from the original PR.

sam-github commented:
@gibfahn I think PR-URL should point to the PR where this *specific* commit was reviewed, and the meta-data should refer to the specific (backport) PR.

So, something like this:

The original PR would be:

```
commit message

PR-URL: original PR (as it is now)
... (metadata as it is now)
```

then it gets backported and PRed, a *new* PR (obviously):

```
commit message

PR-URL: the URL for the PR where the backport was proposed and discussed
Backport-of: the URL to the original PR that needed to be backported
... metadata of *the backport* (not a copy of the original)
```

I changed `Backport` to `Backport-of` because your question made me realize calling the original PR a "backport" is strange and confusing.

If its way too much work, maybe it can't happen, but landing backports, which are rewritten PRs, with metadata claiming they have been reviewed by people who may not in fact have reviewed the rewrite seems misleading.

MylesBorins commented:
We have so far been using the original PR for the PR-URL and including the backport PR as a `Refs:`

This has not been consistent though. I'm up for whatever re: labelling as long as it is consistent and documented

gibfahn commented:
>I changed `Backport` to `Backport-of` because your question made me realize calling the original PR a "backport" is strange and confusing.

@sam-github yeah, that was my issue. I'm fine with using `Backport-of:` for the original PR to master, it makes sense that `PR-URL:`` should always point to the actual PR that added this specific commit.

The `Reviewed-By:` field is (in my experience) already redone to only contain the reviewers of the backport PR.

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+The "Current" release line (currently v7.x) is much more lenient than the LTS
+release lines in what can be landed. Our LTS release lines
+(currently v4.x and v6.x) require that commits live in a Current release for at
+least 2 weeks before backporting. Please see the [`LTS Plan`][] for more
```
Is there a reason you used 

```
[`LTS Plan`]
```

rather than:

```
[LTS Plan]
```

?

gibfahn replied:
As I understand it, the two week rule doesn't apply to doc fixes and infrastructure changes (things that don't change what we ship). This is probably a question for @nodejs/lts, but either way we should probably be explicit here.

Refs: https://github.com/nodejs/node/pull/11351#issuecomment-280226865 and https://github.com/nodejs/build/issues/613#issuecomment-278556390

cc/ @bnoordhuis @MylesBorins @misterdjules

jasnell replied:
The length of time is not set in stone as is based on the potential impact. A doc change has no impact on runtime stability so those, as long as they are relevant to the version, can pretty much land any time. A code change, however, should go through at least one `current` release cycle to ensure that there are no regressions reported. For changes to more sensitive parts of the code, the length of time may be longer. Anything that is semver-minor will typically be batched up and held much longer than semver-patch so that we have less frequency between LTS minors.

evanlucas replied:
Based on what James said, I think we should remove this part. How does one know what sensitive parts of the codebase are when submitting a backport PR? It's hard to document something that isn't concrete imo.

gibfahn replied:
It's definitely a grey area, but people do talk about the two week rule, so I think we should try and document it somehow, even if we have to say that it's quite subjective.

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+is reviewed. When each commit is landed, the new reviewers and the new PR-URL
+should be used.
+
+[`LTS Plan`]: https://github.com/nodejs/LTS#lts-plan
```
Is there a reason you used 

```
[`LTS Plan`]
```

rather than:

```
[LTS Plan]
```

? It looks a bit odd to me.

gibfahn commented:
I also had another question, if @nodejs/lts decides what gets backported to `LTS`, and @nodejs/release do the releases, who decides what gets backported to `current`?

jasnell commented:
Essentially anything sitting in master that is not semver-major or explicitly marked dont-land for the `current` should land in `current`. The person doing the release selects the set. If a PR needs to be backported, then either the release person will do it or will ask someone to do it, otherwise it won't make it in.

rvagg commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+# How to backport a Pull Request to a Release Line
```
'Backport' to make it consistently titlecase eh?

rvagg commented on doc/guides/backporting-to-release-lines.md:
````diff
@@ -0,0 +1,124 @@
+```
+
+* Make the required changes to remove the conflicts, and then commit the
+  changes. That can be done with `git commit`.
````
don't forget `git add` to flag the file as a conflict resolve

MylesBorins commented:
Can we please add to this guide that backport PRs should allow force pushes from collaborators. This is the easiest way to avoid constant rebases

evanlucas commented:
@gibfahn @rvagg @MylesBorins mind taking one more look please? All of yalls concerns should have been addressed. Thanks!

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+## What needs to be backported?
+
+If a cherry-pick from master does not land cleanly on a staging branch, the
+releaser will mark the pull request with a particular label for that release
```
should we be more specific regarding the label here?

jasnell replied:
Yes please, specificity is important

gibfahn replied:
@MylesBorins I think if we see this (at least initially) as a basic guide for people raising backport PRs, then we don't need to specify this yet (all they care about is the comment).

MylesBorins replied:
@gibfahn should we specify the label?

gibfahn replied:
@MylesBorins I think that should probably be done in https://github.com/nodejs/node/pull/12431

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+
+## What can be backported?
+
+The "Current" release line (currently v7.x) is much more lenient than the LTS
```
s/(currently v7.x)/(v7.x as of Feb 2017)

jasnell replied:
Please drop the `v` in the version.

mhdawson replied:
maybe just drop the (currently v7.x) as it is quickly going to be out of date.

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+The "Current" release line (currently v7.x) is much more lenient than the LTS
+release lines in what can be landed. Our LTS release lines
+(currently v4.x and v6.x) require that commits live in a Current release for at
+least 2 weeks before backporting. Please see the [LTS Plan][] for more
```
this is a bit confusing in the way it is selectively using / switching backporting + cherry-picking.

Perhaps 

> Our LTS release lines (v4.x and v6.x as of Feb 2017) require that commits mature in a Current release for at least 2 weeks before they can be landed on staging. If the commit can not be cherry-picked from master a manual backport will need to be submitted. Please see the [LTS Plan][] for more information.

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+error: could not apply 773cdc3... <commit title>
+hint: after resolving the conflicts, mark the corrected paths
+hint: with 'git add <paths>' or 'git rm <paths>'
+hint: and commit the result with 'git commit'
```
isn't the correct thing here to do `git cherry-pick continue`?

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+
+* Make the required changes to remove the conflicts, add the files to the index
+  using `git add`, and then commit the changes. That can be done with
+  `git commit`.
```
I believe this should be `git cherry-pick --continue`

evanlucas replied:
interesting, I've always just fixed it up and then committed manually

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+* Make the required changes to remove the conflicts, add the files to the index
+  using `git add`, and then commit the changes. That can be done with
+  `git commit`.
+* The commit message should be as close as possible to the commit message on the
```
AFAIK I have never changed the original commit message. You can add content to the end of it... but the original message should not chage

evanlucas replied:
I think there are definitely cases where the commit message should change. Especially when the original commit relies on something that is breaking.

MylesBorins replied:
To be more specific, the original message should be preserved, that does not preclude adding additional information

MylesBorins replied:
To be more specific, the original message should be preserved, that does not preclude adding additional information

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+  in the master branch.
+* Push the changes to your fork and open a pull request.
+* Be sure to target the `v7.x-staging` branch in the pull request.
+* When landing a backport commit, please include the PR-URL from the original
```
/cc @nodejs/lts is everyone on board with this meta data change.

For the record we have not been doing this at all. I for one would like to see this come up at an LTS meeting before we commit to taking on this process. If not in a meeting I would at least like to head from @jasnell @rvagg @mhdawson and @Fishrock123 in this comment thread

jasnell replied:
Including the link to the original PR is fine but if this a backport-PR then a different metadata label should be used to avoid accidentally confusing tooling. For instance:

```
Original-PR-URL: https://... (original PR)
PR-URL: https://... (backport PR)
```

gibfahn replied:
@jasnell After the previous discussion in this PR, the consensus for the name (so far) was `Backport-of: <origin PR-URL>`, it's on the next line.

jasnell replied:
Thanks @gibfahn ... missed that earlier discussion

evanlucas replied:
@jasnell that being said, are you alright with that approach?

jasnell replied:
Yes, that definitely works.

mhdawson replied:
I good with the approach I think including the link to the original PR in Backport-of: is the right thing to do.

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+  * Ex. `(v4.x backport) process: improve performance of nextTick`
+* Please include the text `Backport of #<pull request number>` in the
+  pull request description. This will link to the original pull request.
+* Please check the checkbox labelled "Allow edits from maintainers".
```
a++

MylesBorins commented:
Hey All

So I'm doing another batch backport and I think that we really need to take a moment to re-evaluate our process and distribute the workload.

I'd like to suggest that we make a backporting team, potentially a sub committee of the LTS group. This team should hold a couple meetings to properly figure out how to distribute the workload. Part of this will be re-imagining the current labelling process and meta data process.

with that being said I think it makes sense for us to block movement on this doc for now. Then use this doc as a starting point for the backporting team.

The number 1 thing I would like to see is a doc being drafted by / agreed upon by the people who are going to volunteer to do the work.

jasnell commented:
Big +1 to a backporting team organized under the LTS WG. That's the right place for this and the process needs to be determined by the people actually doing the backporting. Definite +1 on blocking this PR for now.

sam-github commented:
If this doc describes current practice, I'd prefer to land it, because even as we reconsider an approach, people still need to know what to do to get things backported. If its unlandable, though, c'est la vie.

I'm available to spend time on backporting, and I'm interested in being involved in the working group, and I agree that now that he basic LTS strategy is working so well (at least, I think the LTS releases have been very sucessful!), that its time to put some more energy towards the maintenance process.

gibfahn commented:
>with that being said I think it makes sense for us to block movement on this doc for now.

I agree with everything else, but as @sam-github said I think we should get something landed so that people can refer to it. As new people get involved I assume we'll want to revise the process, but for now we have an existing process and no clear documentation for it. When people ask (e.g. in https://github.com/nodejs/node/pull/11134#issuecomment-284660794) I'd like to be able to refer them to an existing document.

tl;dr let's get something written down and iterate on it.

mhdawson commented:
I also think there is value to have a doc which captures the current process and helps those doing backports while the new team figures out how to make things better.  It also gives them a stating point to make incremental changes if that is easier.

sam-github commented:
The `Backport-of`, which is a new thing, can be removed if this is going to be stripped down to a "how we do it now" doc, and then the backporting WG can discuss improvements to the process afterwards.

sam-github commented:
https://github.com/nodejs/node/commit/8dbd562590047b35b97b1064b6f1726057979093#commitcomment-21189790 <--- Lack of a documented backport process.

@MylesBorins how soon can we get a backporting WG together? Is that WG actually different from the LTS WG? I gues LTS has regular "should we backport this meetings"?

sam-github commented:
https://github.com/nodejs/LTS/issues/184 <-- opened an issue to specifically discuss @MylesBorins' suggestion.

MylesBorins commented:
if we want to land this now we need to modify it to contain only things that are part of our current workflow.

vsemozhetbyt commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+Below are examples of an original commit message and a backport commit message.
+
+In this example, https://github.com/nodejs/node/pull/1234 is the original pull
+request and https://github.com/nodejs/node/pull/5678 is the backport.
```
These URL dummies become real URLs and get to be confusing as absolutely different PRs.

targos replied:
@gibfahn my other comment was because of this. If the PR-URL is correct, then this sentence is a bit confusing because PR 5678 doesn't appear in the example

vsemozhetbyt commented:
I try to use this guide to make my first backport.

1. I use cloned fork as described in the [CONTRIBUTING.md#Step 1: Fork](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-1-fork)
2. When I try `git checkout v7.x-staging` I get the error `error: pathspec 'v7.x-staging' did not match any file(s) known to git.`

Should I try something like `git checkout -b v7.x-staging -t origin/v7.x-staging` ?

targos commented:
> Should I try something like `git checkout -b v7.x-staging -t origin/v7.x-staging`?

Yes. This is correct.
I think you get this error when the branch exists in more than one of your remotes.

sam-github commented:
For most of us, this would be `git checkout upstream/v7.x-staging -t`, your origin is likely your fork on github, and its v7.x-staging branch won't get updated when the v7.x-staging on nodejs/node does (at least, not automatically).

vsemozhetbyt commented:
@sam-github Does the `git pull -r upstream v7.x-staging` proposed in the guide fix this?

sam-github commented:
fix what?

gibfahn commented:
@vsemozhetbyt I don't think it will, we could either do:

```bash
git checkout -b v7.x-staging
git pull -r upstream v7.x-staging
```

or just do:

```bash
git checkout upstream/v7.x-staging -t
git pull
```

Of course the `git checkout` will fail if you already have a branch, but I'm not sure we need to cover that case.

vsemozhetbyt commented:
@sam-github 
`origin/v7.x-staging` being out of sync with `upstream/v7.x-staging`

sam-github commented:
No, will not fix that, and its not even a problem, IMO. I personally whenever I fork a repo in github immediately delete every branch in my fork. github forks have no method of auto-syncing against upstream, its a strangely missing feature, so I just don't bother. My fork, origin, has only my branches, nodejs/node has only its. I make a concession for master, I do a `git fetch; git push origin upstream/master:master` regularly to keep my fork up to date.

But anyhow, long and short is that when you create a working branch from a branch that is maintained and updated in nodejs/node, you need to fetch the upstream, then branch from upstream.

vsemozhetbyt commented:
So `git checkout v7.x-staging` supposes I've already created local `v7.x-staging` branch earlier?

gibfahn commented:
@sam-github I go one step further and rename `fork/master` to `oldmaster`, that way I never accidentally end up on an old master cause I forgot to update (I documented how I do it [here](https://gist.github.com/gibfahn/a7f35d6d6095d7b9fb860666f1b6a172)).

@vsemozhetbyt not quite, if you do `git checkout v7.x-staging` and you have no local branches by that name, and only one remote branch, it will automatically do `git checkout upstream/v7.x-staging -t`. However if there are two remote branches with that name Git can't tell which branch you meant so it errors.

vsemozhetbyt commented:
@gibfahn But the  [CONTRIBUTING.md#Step 1: Fork](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-1-fork) sets two remotes (origin (fork) and upstream) as a default start point. Should not this be taken into consideration in the guide for beginners in backporting?

gibfahn commented:
@vsemozhetbyt to be clear, I'm saying that this is why `git checkout v7.x-staging` normally works. For this guide I think it should say:

```bash
git checkout upstream/v7.x-staging -t
git pull
```

(unless someone has a better git idea).

Qard commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+pad while preparing a release. The branch name is formatted as follows:
+`vN.x-staging` where `N` is the major release number.
+
+### Active staging branches
```
Might be best to avoid embedding "current" versions in here as they won't stay current long and this doc would need to get updated constantly.

vsemozhetbyt commented:
It seems this does not always work:
```
git checkout -b vN.x-staging
git pull -r upstream vN.x-staging
```
I was trying to create a backport for the v4 and I've got this error:
<details>
<summary>Output (click me):</summary>

```
$ git checkout -b v4.x-staging

$ git pull -r upstream v4.x-staging

remote: Counting objects: 94, done.
remote: Compressing objects: 100% (52/52), done.
remote: Total 94 (delta 39), reused 18 (delta 18), pack-reused 24
Unpacking objects: 100% (94/94), done.
From https://github.com/nodejs/node
 * branch                  v4.x-staging -> FETCH_HEAD
   ba0b8388b1..67f9f85aed  v4.x-staging -> upstream/v4.x-staging
First, rewinding head to replay your work on top of it...
Applying: Working on v5.0.0
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M       src/node_version.h
Falling back to patching base and 3-way merge...
Auto-merging src/node_version.h
CONFLICT (content): Merge conflict in src/node_version.h
Patch failed at 0001 Working on v5.0.0
The copy of the patch that failed is found in: .git/rebase-apply/patch

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
```
</details>
If I get it right, `git checkout -b vN.x-staging` creates a branch based on the current master, so rebasing into an older version branch would cause many conflicts.

While this worked OK:
```
git checkout upstream/v4.x-staging -t
git pull
```

gibfahn commented:
@vsemozhetbyt good point, so we could do:

```bash
git branch v7.x-staging upstream/v7.x-staging    # Fails if you already have a v7.x-staging
git checkout v7.x-staging
git reset --hard upstream/v7.x-staging
```

gibfahn commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,124 @@
+  metadata from the original commit should be removed. If a backport is
+  required, it should go through the same review steps as a commit landing
+  in the master branch.
+* Push the changes to your fork and open a pull request.
```
There should probably be a:

> * Make sure `make -j4 test` still passes.

above this line, see https://github.com/nodejs/node/pull/11797#discussion_r105559111

gibfahn commented:
@evanlucas I think this is the current list of changes it'd take to reduce this to what we actually do today. If we can do that we can get this landed as a guide for people who've been asked to do backports, and work improving the process separately.

As far as the PR metadata goes, I assume that, as for PRs to master, it's currently modified by whoever lands the PR, so we probably don't need to be too specific.

<details><summary> Diff of changes </summary>


<pre lang=diff>
diff --git a/doc/guides/backporting-to-release-lines.md b/doc/guides/backporting-to-release-lines.md
index 5f48c002dd..a9cb860023 100644
--- a/doc/guides/backporting-to-release-lines.md
+++ b/doc/guides/backporting-to-release-lines.md
@@ -47,8 +47,9 @@ that with the staging branch for the targeted release line.
 # the origin remote points to your fork, and the upstream remote points
 # to git://github.com/nodejs/node
 cd $NODE_DIR
+git branch v7.x-staging upstream/v7.x-staging    # Fails if you already have a v7.x-staging
 git checkout v7.x-staging
-git pull -r upstream v7.x-staging
+git reset --hard upstream/v7.x-staging
 # We want to backport pr #10157
 git checkout -b backport-10157-to-v7.x
 ```
@@ -74,11 +75,9 @@ hint: and commit the result with 'git commit'
   metadata from the original commit should be removed. If a backport is
   required, it should go through the same review steps as a commit landing
   in the master branch.
+* Make sure `make -j4 test` passes.
 * Push the changes to your fork and open a pull request.
 * Be sure to target the `v7.x-staging` branch in the pull request.
-* When landing a backport commit, please include the PR-URL from the original
-  pull request as `Backport-of: <origin PR-URL>`. The `Backport-of` line should
-  be directly after the `PR-URL` line in the metadata.
 
 Below are examples of an original commit message and a backport commit message.
 
@@ -103,7 +102,6 @@ lib: make something faster
 Switch to using String#repeat to improve performance.
 
 PR-URL: https://github.com/nodejs/node/pull/5678
-Backport-of: https://github.com/nodejs/node/pull/1234
 ```
 
 ### Helpful Hints
@@ -111,8 +109,6 @@ Backport-of: https://github.com/nodejs/node/pull/1234
 * Please include the backport target in the pull request title in the following
   format: `(v7.x backport) <commit title>`
   * Ex. `(v4.x backport) process: improve performance of nextTick`
-* Please include the text `Backport of #<pull request number>` in the
-  pull request description. This will link to the original pull request.
 * Please check the checkbox labelled "Allow edits from maintainers".
   This is the easiest way to to avoid constant rebases.
 
</pre>

</details>

gibfahn commented:
@MylesBorins the two issues you raised (the `git pull` and the `Backport-Of:` line) have now been fixed and removed respectively. Does this LGTY now? If we have this documenting what we currently do, we can update it as we refine/change the process.

@rvagg LGTY now?

gibfahn commented on doc/guides/backporting-to-release-lines.md:
````diff
@@ -0,0 +1,122 @@
+```
+
+* Make the required changes to remove the conflicts, add the files to the index
+  using `git add`, and then commit the changes. That can be done with
````
Maybe delete `. That can be done`, otherwise it seems like maybe `git cherry-pick continue` is just one way to commit the changes, and we don't want people defaulting to `git commit`.

gibfahn commented:
ping @MylesBorins

MylesBorins commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,122 @@
+
+Switch to using String#repeat to improve performance.
+
+PR-URL: https://github.com/nodejs/node/pull/5678
```
This is not accurate. we maintain the original PR-URL

evanlucas commented:
@rvagg mind taking a quick look? Thanks!

targos commented on doc/guides/backporting-to-release-lines.md:
```diff
@@ -0,0 +1,121 @@
+
+Switch to using String#repeat to improve performance.
+
+PR-URL: https://github.com/nodejs/node/pull/1234
```
5678

gibfahn replied:
@targos see https://github.com/nodejs/node/pull/11099#discussion_r108266680

evanlucas commented:
ping @rvagg

gibfahn commented:
In the [backporting meeting](https://github.com/nodejs/LTS/issues/191) we decided that it would be easiest (at least for now) if people raising Backport PRs didn't have to modify the commit message, so I've added a commit to remove that info.

cc/ @MylesBorins @sam-github @addaleax @mhdawson

gibfahn commented:
Does anyone have any objection to this landing? Any further issues can be looked at later, and it would be really good to have something to refer people to.

sam-github commented:
LGTM, still. Get it in, iterate on it as we go.

gibfahn commented:
Landed in https://github.com/nodejs/node/commit/ece9ab55067c7d6ca20b8295eb0c1f2451e8729c
